samza-core/src/main/java/org/apache/samza/zk/ZkBarrier.java

```diff
@@ -0,0 +1,214 @@
+ *  |   |- barrier_participants/
+ *  |   |   |- {id1}
+ *  |   |   |- {id2}
+ *  |   |   |-  ...
+ */
+public class ZkBarrier {
+  private final static Logger LOG = LoggerFactory.getLogger(ZkBarrier.class);
+  private static final String BARRIER_PARTICIPANTS = "/barrier_participants";
+  private static final String BARRIER_STATE = "/barrier_state";
+  private final ZkUtils zkUtils;
+  private final BarrierKeyBuilder keyBuilder;
+
+  public enum State {
+    TIMED_OUT, DONE
+  }
+
+  private ZkBarrierListener barrierListener;
+
+  public ZkBarrier(String barrierRoot, ZkUtils zkUtils, ZkBarrierListener barrierListener) {
+    if (zkUtils == null) {
+      throw new RuntimeException("Cannot operate ZkBarrierForVersionUpgrade without ZkUtils.");
+    }
+    this.zkUtils = zkUtils;
+    this.keyBuilder = new BarrierKeyBuilder(barrierRoot);
+    this.barrierListener = barrierListener;
+  }
+
+  /**
+   * Creates a shared barrier sub-tree in ZK
+   *
+   * @param version Version associated with the Barrier
+   * @param participants List of expected participated for this barrier to complete
+   */
+  public void create(final String version, List<String> participants) {
+    String barrierRoot = keyBuilder.getBarrierRoot();
+    zkUtils.makeSurePersistentPathsExists(new String[]{
+        barrierRoot,
+        keyBuilder.getBarrierPath(version),
+        keyBuilder.getBarrierParticipantsPath(version),
+        keyBuilder.getBarrierStatePath(version)});
+
+    // subscribe for participant's list changes
+    String barrierParticipantsPath = keyBuilder.getBarrierParticipantsPath(version);
+    LOG.info("Subscribing for child changes at " + barrierParticipantsPath);
+    zkUtils.getZkClient().subscribeChildChanges(barrierParticipantsPath, new ZkBarrierChangeHandler(version, participants));
+
+    if (barrierListener != null) {
+      barrierListener.onBarrierCreated(version);
+    }
+  }
+
+  /**
+   * Joins a shared barrier by registering under the barrier sub-tree in ZK
+   *
+   * @param version Version associated with the Barrier
+   * @param participantId Identifier of the participant
+   */
+  public void join(String version, String participantId) {
+    String barrierDonePath = keyBuilder.getBarrierStatePath(version);
+    zkUtils.getZkClient().subscribeDataChanges(barrierDonePath, new ZkBarrierReachedHandler(barrierDonePath, version));
+
+    // TODO: Handle ZkNodeExistsException
+    zkUtils.getZkClient().createPersistent(
+        String.format("%s/%s", keyBuilder.getBarrierParticipantsPath(version), participantId));
+  }
+
+  /**
+   * Expires the barrier version by marking it as TIMED_OUT
+   *
+   * @param version Version associated with the Barrier
+   */
+  public void expire(String version) {
+    zkUtils.getZkClient().writeData(
+        keyBuilder.getBarrierStatePath(version),
+        State.TIMED_OUT);
+
+  }
+  /**
+   * Listener for changes to the list of participants. It is meant to be subscribed only by the creator of the barrier
```

Comment:
We could make the creator register it as an anonymous class. That way, non-creators don't have an option of even using this class. But in reality, it doesn't prevent anyone from subscribing to any of the znode paths. 

Since this is internal to this implementation of the barrier, I suspect the probability of non-creators becoming listeners is really low.